qemu/tests/slof_boot.py

```diff
@@ -0,0 +1,117 @@
+    :param env: Dictionary with test environment.
+    """
+
+    def _get_pci_bridge_addr(id):
+        dev_info_list = re.findall(r'\s+-device pci-bridge,\S+\s+',
+                                   vm.qemu_command)
+        for dev_info in dev_info_list:
+            if ('id=' + id) in dev_info:
+                return re.search(r'(addr)=(\w+)', dev_info).group(2)
+
+    def _verify_boot_status(boot_dev):
+        dev_params = params.object_params(boot_dev)
+        child_addr = dev_params.get('child_addr')
+        sub_child_addr = dev_params.get('sub_child_addr', None)
+        parent_bus = dev_params.get('parent_bus')
+        child_bus = dev_params.get('child_bus')
+        if child_bus == 'pci-bridge':
+            pci_bus_id = params.get('pci_bus_image1', None)
+            child_addr = _get_pci_bridge_addr(pci_bus_id)
+        if sub_child_addr:
+            verify_info = ('Verify whether booting from specified %s device (%s@%s).'
+                           % (boot_dev, child_addr, sub_child_addr))
+            fail_info = ('Failed to boot from %s device (%s@%s).'
+                         % (boot_dev, child_addr, sub_child_addr))
+            ret_info = ('Boot from %s device (%s@%s) successfully.'
+                        % (boot_dev, child_addr, sub_child_addr))
+        else:
+            verify_info = ('Verify whether booting from specified %s device (@%s).'
+                           % (boot_dev, child_addr))
+            fail_info = ('Failed to boot from %s device (@%s).'
+                         % (boot_dev, child_addr))
+            ret_info = ('Boot from %s device (@%s) successfully.'
+                        % (boot_dev, child_addr))
+
+        logging.info(verify_info)
+
+        if not slof.verify_boot_device(vm, parent_bus, child_bus,
+                                       child_addr, sub_child_addr):
+            test.fail(fail_info)
+        else:
+            logging.info(ret_info)
+
+    timeout = float(params.get("login_timeout", 240))
+    vm = env.get_vm(params['main_vm'])
+
+    if not slof.wait_for_loaded(vm)[0]:
+        test.fail('SLOF boot failed.')
+
+    error_context.context("Check the output of SLOF.", logging.info)
+    ret = slof.check_error(vm)
+    if ret:
+        test.fail("Found error info: %s ." % ret)
+    error_context.context("No any error info", logging.info)
+
+    _verify_boot_status(params['boot_dev_type'])
+
+    error_context.context("Try to log into guest '%s'." % vm.name,
+                          logging.info)
+    session = vm.wait_for_login(timeout=timeout)
+    error_context.context("log into guest '%s' successfully." % vm.name,
+                          logging.info)
+
+    error_context.context("Try to ping external host.", logging.info)
+    extra_host_ip = utils_net.get_host_ip_address(params)
+    s, o = session.cmd_status_output('ping %s -c 5' % extra_host_ip)
+    logging.debug(o)
+    if s:
+        test.fail("Failed to ping external host.")
+    else:
```

Comment:
`else:` can be omitted and unintended the next line

Reply:
Updated , please check. thanks.